ee/modules/021-cni-cilium/hooks/ee/egressgateways_discovery.go

```diff
@@ -0,0 +1,519 @@
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube/object_patch"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8type "k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/strings/slices"
+
+	eeCrd "github.com/deckhouse/deckhouse/go_lib/cilium/apis/v1alpha1"
+)
+
+const (
+	cniNamespace      = "d8-cni-cilium"
+	memberLabelKey    = "egress-gateway.network.deckhouse.io/member"
+	masterLabelKey    = "egress-gateway.network.deckhouse.io/master-node-name"
+	masterLabelPrefix = "egress-gateway.network.deckhouse.io/master-"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 10},
+	Queue:        "/modules/cni-cilium/egress-discovery",
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:       "egressgateways",
+			ApiVersion: "network.deckhouse.io/v1alpha1",
+			Kind:       "EgressGateway",
+			FilterFunc: applyEgressGatewayFilter,
+		},
+		{
+			Name:       "egressgatewayinstances",
+			ApiVersion: "network.deckhouse.io/v1alpha1",
+			Kind:       "EgressGateway",
+			FilterFunc: applyEgressGatewayInstanceFilter,
+		},
+		{
+			Name:       "nodes",
+			ApiVersion: "v1",
+			Kind:       "Node",
+			FilterFunc: applyNodeFilter,
+		},
+		{
+			Name:       "cilium_agent_pods",
+			ApiVersion: "v1",
+			Kind:       "Pod",
+			NamespaceSelector: &types.NamespaceSelector{
+				NameSelector: &types.NameSelector{
+					MatchNames: []string{cniNamespace},
+				},
+			},
+			LabelSelector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": "egress-gateway-agent",
+				},
+			},
+			FilterFunc: applyPodFilter,
+		},
+	},
+}, handleEgressGateways)
+
+func applyEgressGatewayFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var eg eeCrd.EgressGateway
+
+	err := sdk.FromUnstructured(obj, &eg)
+	if err != nil {
+		return nil, err
+	}
+
+	return EgressGatewayInfo{
+		Generation:   eg.GetGeneration(),
+		Name:         eg.Name,
+		UID:          eg.GetUID(),
+		NodeSelector: eg.Spec.NodeSelector,
+		SourceIP: EgressGatewaySourceIP{
+			Mode: string(eg.Spec.SourceIP.Mode),
+			VirtualIPAddress: VirtualIPAddress{
+				IP:               eg.Spec.SourceIP.VirtualIPAddress.IP,
+				RoutingTableName: eg.Spec.SourceIP.VirtualIPAddress.RoutingTableName,
+			},
+			PrimaryIPFromEgressGatewayNodeInterface: PrimaryIPFromEgressGatewayNodeInterface{
+				InterfaceName: eg.Spec.SourceIP.PrimaryIPFromEgressGatewayNodeInterfaceSpec.InterfaceName,
+			},
+		},
+	}, nil
+}
+
+func applyEgressGatewayInstanceFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var egi eeCrd.EgressGatewayInstance
+
+	err := sdk.FromUnstructured(obj, &egi)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if finalizer is empty and remove that EgressGatewayInstance
+	var toRemove bool
+	if len(egi.Finalizers) == 0 {
```

Comment:
Wrong logic, we just DELETE the resource and agent wipe the finalizer. Then kube-apiserver will finally delete the resource.